src/sage/combinat/symmetric_group_algebra.py

```diff
@@ -2068,8 +2068,152 @@ def dft(self, form=None, mult='l2r'):
+            sage: U = GF5S3._dft_unitary()
+            sage: U*U.H == identity_matrix(GF5S3.group().cardinality())
+            True
+        """
+        from sage.matrix.special import diagonal_matrix
+        from sage.misc.functional import sqrt
+        from sage.misc.flatten import flatten
+
+        G = self.group()
+        F = self.base_ring()
+        assert F.is_field()
+
+        if F.characteristic() == 0:
+            data = {}
+            for partition in Partitions(G.degree()):
+                specht_module = self.specht_module(partition)
+                rho = specht_module.representation_matrix
+                P = sum(rho(g) * rho(g).conjugate().transpose() for g in G) / G.cardinality()
+                d, L = P.eigenmatrix_left()
+                data[partition] = (specht_module, P, d, L)
+
+            def all_roots_field():
+                required_square_roots = []
+                for partition in Partitions(G.degree()):
+                    specht_module, P, d, L = data[partition]
+                    required_square_roots += [specht_module.dimension(), G.cardinality()] + d.diagonal()
+                required_square_roots = flatten([[QQ(q).numerator(), QQ(q).denominator()] if q in QQ else q for q in required_square_roots])
+                K = F
+                for n in set(required_square_roots):
+                    R = PolynomialRing(K, 'x')
+                    x = R.gen()
+                    if n.is_rational() and (x**2 - n).is_irreducible():
+                        gen_name = "sqrt"+str(n).replace("/","over")
+                        K = K.extension(sqrt(n).minpoly(),names=gen_name)
+                    if not n.is_rational() and sqrt(n).minpoly().is_irreducible():
+                        gen_name = "deg" + str(n.minpoly().degree()) + "index" + str(list(set(required_square_roots)).index(n))
+                        K = K.extension(sqrt(n).minpoly(),names=gen_name)
+                return K
+
+            def unitary_change_of_basis(partition, K):
+                specht_module, P, d, L = data[partition]
+                return L.inverse() * diagonal_matrix([sqrt(K(a)) for a in d.diagonal()]) * L
+            K = all_roots_field()
+            unitary_data = {partition: unitary_change_of_basis(partition, K) for partition in Partitions(G.degree())}
+
+            def hat(f, partition):
+                specht_module, P, d, L = data[partition]
+                rho = self.specht_module(partition).representation_matrix
+                Q = unitary_data[partition]
+                unitary_factor = specht_module.dimension() / G.cardinality()
+                sqrt_unitary_factor = sqrt(K(unitary_factor))
+                return sqrt_unitary_factor * sum(f(g) * Q.inverse() * rho(g) * Q for g in G if f(g))
+
+            delta = lambda s: lambda t: 1 if t == s else 0
+            return matrix(K, [flatten([hat(delta(g), partition).list() for partition in Partitions(G.degree())]) for g in G]).transpose()
```

Comment:
```suggestion
            def hat(g, partition):
                specht_module, P, d, L = data[partition]
                rho = self.specht_module(partition).representation_matrix
                Q = unitary_data[partition]
                unitary_factor = specht_module.dimension() / G.cardinality()
                sqrt_unitary_factor = sqrt(K(unitary_factor))
                return sqrt_unitary_factor * Q.inverse() * rho(g) * Q

            PGdeg = Partitions(G.degree())
            return matrix(K, [flatten([hat(g, partition).list() for partition in PGdeg]) for g in G]).transpose()
```
I think this is equivalent since `delta` is just getting the summand `g` .

Although you probably can write this to get rid of that transpose and the `flatten`... The partitions parent `PGdeg` can also be created earlier (and reused here).

Reply:
Yes, this was written to be general and obtain the Fourier coefficient for any function f. Since we only need the summand `g`, this is simpler.